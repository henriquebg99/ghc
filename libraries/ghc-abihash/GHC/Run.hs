module GHC.Run (runGhcWithAbiHashes, runGhcTWithAbiHashes, initGhcMonadWithAbiHashes) where

import GHC.AbiHashes
import GHC.Utils.Exception
import GHC

-- | The same as 'runGhc', but passes in the ABI hashes of
-- GHC and all its dependencies
-- Use this function when you don't have any additional requirements
-- for linking consistency in your GHC-API program, i.e. if you don't
-- have your own runtime loading/plugin system other than GHCs plugin
-- and TH support.
-- If you do have your own plugin system on top of GHCs, it is probably a good
-- idea to add the ABI hashes of the plugin interface and all its dependencies
-- to 'ghcAbiHashes' and use 'GHC.runGhc' directly
-- See Note [Loader Consistency Checks] for more details
runGhcWithAbiHashes
 :: Maybe FilePath  -- ^ See argument to 'initGhcMonad'.
 -> Ghc a           -- ^ The action to perform.
 -> IO a
runGhcWithAbiHashes = runGhc ghcAbiHashes

-- | The same as 'runGhcT', but passes in the ABI hashes of
-- GHC and all its dependencies
-- Use this function when you don't have any additional requirements
-- for linking consistency in your GHC-API program, i.e. if you don't
-- have your own runtime loading/plugin system other than GHCs plugin
-- and TH support.
-- If you do have your own plugin system on top of GHCs, it is probably a good
-- idea to add the ABI hashes of the plugin interface and all its dependencies
-- to 'ghcAbiHashes' and use 'GHC.runGhc' directly
-- See Note [Loader Consistency Checks] for more details
runGhcTWithAbiHashes :: ExceptionMonad m
        => Maybe FilePath  -- ^ See argument to 'initGhcMonad'.
        -> GhcT m a        -- ^ The action to perform.
        -> m a
runGhcTWithAbiHashes = runGhcT ghcAbiHashes

-- | The same as 'initGhcMonad', but passes in the ABI hashes of
-- GHC and all its dependencies
-- Use this function when you don't have any additional requirements
-- for linking consistency in your GHC-API program, i.e. if you don't
-- have your own runtime loading/plugin system other than GHCs plugin
-- and TH support.
-- If you do have your own plugin system on top of GHCs, it is probably a good
-- idea to add the ABI hashes of the plugin interface and all its dependencies
-- to 'ghcAbiHashes' and use 'GHC.runGhc' directly
-- See Note [Loader Consistency Checks] for more details
initGhcMonadWithAbiHashes :: GhcMonad m => Maybe FilePath -> m ()
initGhcMonadWithAbiHashes = initGhcMonad ghcAbiHashes
