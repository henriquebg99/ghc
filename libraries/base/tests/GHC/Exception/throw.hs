module Main where

import ThrowTestLib
import GHC.Exception

main = runThrowTest throw
