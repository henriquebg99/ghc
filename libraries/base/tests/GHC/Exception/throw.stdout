"=== Test backtraces ==="
"Backtrace mechanisms []:"
[]
"Backtrace mechanisms [HasCallStackBacktraceMech]:"
[[("collectHasCallStackBacktrace",SrcLoc {srcLocPackage = "base", srcLocModule = "GHC.Exception.Backtrace", srcLocFile = "libraries/base/GHC/Exception/Backtrace.hs", srcLocStartLine = 109, srcLocStartCol = 52, srcLocEndLine = 109, srcLocEndCol = 80}),("collectBacktraces'",SrcLoc {srcLocPackage = "base", srcLocModule = "GHC.Exception.Backtrace", srcLocFile = "libraries/base/GHC/Exception/Backtrace.hs", srcLocStartLine = 102, srcLocStartCol = 27, srcLocEndLine = 102, srcLocEndCol = 45}),("collectBacktraces",SrcLoc {srcLocPackage = "base", srcLocModule = "GHC.Exception", srcLocFile = "libraries/base/GHC/Exception.hs", srcLocStartLine = 66, srcLocStartCol = 30, srcLocEndLine = 66, srcLocEndCol = 47}),("throw",SrcLoc {srcLocPackage = "main", srcLocModule = "Main", srcLocFile = "throw.hs", srcLocStartLine = 6, srcLocStartCol = 21, srcLocEndLine = 6, srcLocEndCol = 26})]]
"Backtrace mechanisms [CostCenterBacktraceMech]:"
[["Main.main (throw.hs:6:1-25)","ThrowTestLib.runThrowTest (ThrowTestLib.hs:(18,1)-(34,20))","ThrowTestLib.printBacktrace (ThrowTestLib.hs:(37,1)-(48,9))","ThrowTestLib.produceDeepStack (ThrowTestLib.hs:(52,1)-(63,53))","ThrowTestLib.produceDeepStack.getDeepStackCase (ThrowTestLib.hs:(62,5)-(63,53))"]]
"Backtrace mechanisms [IPEBacktraceMech]:"
[[StackEntry {functionName = "runThrowTest", moduleName = "ThrowTestLib", srcLoc = "ThrowTestLib.hs:26:18-30", closureType = 53},StackEntry {functionName = "printBacktrace", moduleName = "ThrowTestLib", srcLoc = "ThrowTestLib.hs:(41,5)-(47,7)", closureType = 53},StackEntry {functionName = "printBacktrace", moduleName = "ThrowTestLib", srcLoc = "ThrowTestLib.hs:42:48", closureType = 53}]]
"Backtrace mechanisms [IPEBacktraceMech,CostCenterBacktraceMech,HasCallStackBacktraceMech]:"
[[StackEntry {functionName = "runThrowTest", moduleName = "ThrowTestLib", srcLoc = "ThrowTestLib.hs:28:5-17", closureType = 53},StackEntry {functionName = "printBacktrace", moduleName = "ThrowTestLib", srcLoc = "ThrowTestLib.hs:(41,5)-(47,7)", closureType = 53},StackEntry {functionName = "printBacktrace", moduleName = "ThrowTestLib", srcLoc = "ThrowTestLib.hs:42:48", closureType = 53}],["Main.main (throw.hs:6:1-25)","ThrowTestLib.runThrowTest (ThrowTestLib.hs:(18,1)-(34,20))","ThrowTestLib.printBacktrace (ThrowTestLib.hs:(37,1)-(48,9))","ThrowTestLib.produceDeepStack (ThrowTestLib.hs:(52,1)-(63,53))","ThrowTestLib.produceDeepStack.getDeepStackCase (ThrowTestLib.hs:(62,5)-(63,53))"],[("collectHasCallStackBacktrace",SrcLoc {srcLocPackage = "base", srcLocModule = "GHC.Exception.Backtrace", srcLocFile = "libraries/base/GHC/Exception/Backtrace.hs", srcLocStartLine = 109, srcLocStartCol = 52, srcLocEndLine = 109, srcLocEndCol = 80}),("collectBacktraces'",SrcLoc {srcLocPackage = "base", srcLocModule = "GHC.Exception.Backtrace", srcLocFile = "libraries/base/GHC/Exception/Backtrace.hs", srcLocStartLine = 102, srcLocStartCol = 27, srcLocEndLine = 102, srcLocEndCol = 45}),("collectBacktraces",SrcLoc {srcLocPackage = "base", srcLocModule = "GHC.Exception", srcLocFile = "libraries/base/GHC/Exception.hs", srcLocStartLine = 66, srcLocStartCol = 30, srcLocEndLine = 66, srcLocEndCol = 47}),("throw",SrcLoc {srcLocPackage = "main", srcLocModule = "Main", srcLocFile = "throw.hs", srcLocStartLine = 6, srcLocStartCol = 21, srcLocEndLine = 6, srcLocEndCol = 26})]]
